Add requested features

hdavid16 opened the pull request:
This PR merges the following individual PR's that address feature requests. I merged them to avoid conflicts when rebasing each time a different PR is merged.

#178 - set background color (conflicts with changing plot size; conflict fixed in this PR). Closes issue #172.
#181 - fix bug in shell_layout. Closes issue #149 
#182 - set plot size. Closes issues #94 #147 #161
#183 - add plot title. Closes issue #107
#184 - make self-loops curved. Closes issue #160
#185 - fix bug in spring_layout. Closes issue #154 

Also fixes the second point in issue #150 by switching the arrows to be triangles.
Also closes issue #95 
Also fixes issue #190

simonschoelly commented:
I see thee are some errors on Julia 1.3. We probably should make the minimum version of this package v1.6.

simonschoelly commented on src/layout.jl:
```diff
@@ -102,11 +102,11 @@ julia> locs_x, locs_y = spring_layout(g)
 function spring_layout(g::AbstractGraph,
-                       locs_x=2*rand(nv(g)).-1.0,
-                       locs_y=2*rand(nv(g)).-1.0;
+                       locs_x_in::Vector{R1}=2*rand(nv(g)).-1.0,
+                       locs_y_in::Vector{R2}=2*rand(nv(g)).-1.0;
```
Should we really restrict ourselves to `Vector` here?

hdavid16 replied:
I used Vector here to make it consistent with the same `args` in `gplot()` (line 119 of plots.jl). Do you think we should relax it here?

simonschoelly commented on src/layout.jl:
```diff
@@ -119,6 +119,10 @@ function spring_layout(g::AbstractGraph,
     force_y = zeros(nvg)
 
+    # Convert locs to float
+    locs_x = float.(locs_x_in)
+    locs_y = float.(locs_y_in)
```
```suggestion
    locs_x = convert(Vector{Float64, locs_x_in})
    locs_y = convert(Vector{Float64, locs_y_in))
```
This way, we do  not allocate a new array, if the inputs are already of type `Vector{Float64}`.

hdavid16 replied:
I didn't know this. Great. thanks.

simonschoelly commented:
I can review this when I find time, but in general I think it is better to keep different things in different PRs - makes it easier to review.

hdavid16 commented:
> I see thee are some errors on Julia 1.3. We probably should make the minimum version of this package v1.6.

I have increased it to 1.6 and removed old compats. I also removed ColorTypes, which doesn't seem to be used anywhere.

etiennedeg reviewed (commented):
Looks good to me otherrwise

hdavid16 commented on src/lines.jl:
```diff
@@ -201,3 +201,49 @@ function curveedge(x1, y1, x2, y2, θ, outangle, d; k=0.5)
+    A = adjacency_matrix(g) #adjacency matrix
+    B = spdiagm(diag(A)) #diagonal matrix (self-loops)
+    A[diagind(A)] .= 0 #set diagonal elements to 0 (remove self-loops)
+    if is_directed(g)
+        g1 = SimpleDiGraph(A)
+        g2 = SimpleDiGraph(B)
+    else
+        g1 = SimpleGraph(A)
+        g2 = SimpleGraph(B)
+    end
+
+    lines, larrows = build_straight_edges(g1, locs_x, locs_y, nodesize, arrowlengthfrac, arrowangleoffset)
+    curves, carrows = build_curved_edges(g2, locs_x, locs_y, nodesize, arrowlengthfrac, arrowangleoffset, outangle)
```
Good idea, I'll put that together.

hdavid16 replied:
@etiennedeg i made the changes you suggested. Looks much more efficient now. Thanks for the idea.

simonschoelly commented on src/layout.jl:
```diff
@@ -102,11 +102,11 @@ julia> locs_x, locs_y = spring_layout(g)
 function spring_layout(g::AbstractGraph,
-                       locs_x=2*rand(nv(g)).-1.0,
-                       locs_y=2*rand(nv(g)).-1.0;
+                       locs_x_in::Vector{R1}=2*rand(nv(g)).-1.0,
+                       locs_y_in::Vector{R2}=2*rand(nv(g)).-1.0;
```
This will break code, that passed in something else than a `Vector` - maybe it is not too big of a problem though. We could also allow at least as `AbstractVector` maybe?

hdavid16 replied:
Hi @simonschoelly, What else can be passed besides a vector?

hdavid16 replied:
Or we could leave this for a future PR where I plan to revisit the whole source code to make it better overall (e.g., remove the confusing all caps kwargs in `gplot`).

etiennedeg replied:
I think we can change it also in `gplot`, in this direction, it is not breaking.

hdavid16 replied:
done

hdavid16 commented:
@simonschoelly @etiennedeg Any updates?

etiennedeg commented:
Sorry about this. I'm not sure how it fix #150. I think changing for triangles is good, but maybe it would be better in another issue. 
![Capture d’écran du 2022-09-21 16-46-00](https://user-images.githubusercontent.com/22714764/191546415-aae618dc-b746-4b03-a0f2-67e50dd0ca48.png)
Simon's review is not addressed. For the remaining, I think we are good.

hdavid16 commented:
@etiennedeg, #150 is fixed by using arrows instead of angled lines because you can control the corner locations of the arrows. When you use angled lines to make an arrow, the actual tip of the arrow depends on the width of the lines.

Is the picture you are showing generated with the new triangle arrows? I can move the triangle arrows to a separate PR if desired where we can work it out further.

In terms of @simonschoelly 's comments, I have answered each of them and am awaiting for responses to my comments.

I'm fine breaking up these PR's if you and Simon prefer if that will truly help them get reviewed and merged faster. But if they'll still take months to review, I'm not sure of the benefit of having multiple PR's (at least from my point as a user, I want to be able to use these different features simultaneously while I wait for the PR to be reviewed). But I can resplit the PR into several smaller ones. Please confirm what you want me to do in this regard.

etiennedeg commented:
I think #150 is mostly an issue when saving to a non-square picture, because there is an inconsistency between absolute coordinates and relative coordinates which give strange distortions. IMO, the position of arrows in a square image was not an issue (at least not for the examples I tested). My picture is from the output of plots in Juno, and it is not generated as a square picture.

etiennedeg commented:
Another issue I see is when there is two arcs in opposite directions, it looks weird. Maybe we can fix this by starting the edge at the midway point between the two corners of the arrow tip?

hdavid16 commented:
> Another issue I see is when there is two arcs in opposite directions, it looks weird. Maybe we can fix this by starting the edge at the midway point between the two corners of the arrow tip?

Great idea. I have added this fix in the last commit. I now check for if a reverse edge exists and then use the midway point on the arrow base for the line.

hdavid16 commented:
@simonschoelly do you want me to split this up into small PRs? This has been sitting here for several months. If I make small PRs would you or someone else review them? Thanks.

etiennedeg commented:
This is pending for too long and this would greatly improve the situation, we need to merge it. 
As it is now, it seems fine.
@simonschoelly @gdalle, if you want to take a look, otherwise I will merge it in a week.

etiennedeg commented:
@hdavid16 Thanks for the PR, and sorry for the excessive delay.

hdavid16 commented:
Thanks @etiennedeg . Might be good to go a new patch or minor release with the added features.

etiennedeg commented:
Yes, I wonder if we can get #193 in the release, but it will depend on Graphs#376. If it takes too long for someone to review it, I will do a release without it.

etiennedeg commented:
@hdavid16 I did a release
